File: ios/SocialBrain/Models/EncryptedRecordEnvelope.swift
import Foundation

enum RecordType: String, Codable, CaseIterable {
    case person, group, groupMembership, relationship, event, eventAttendee, memory, capture, reminder, settings
}

struct HybridRevision: Codable, Comparable, Hashable {
    let counter: Int64
    let deviceID: UUID
    static func < (lhs: HybridRevision, rhs: HybridRevision) -> Bool {
        lhs.counter == rhs.counter ? lhs.deviceID.uuidString < rhs.deviceID.uuidString : lhs.counter < rhs.counter
    }
    var encoded: String { "\(counter):\(deviceID.uuidString.lowercased())" }
}

struct EncryptedRecordEnvelope: Codable, Identifiable, Equatable {
    let id: UUID
    let recordType: RecordType
    let schemaVersion: Int
    let keyVersion: Int
    let revision: HybridRevision
    let deviceID: UUID
    let updatedAtMs: Int64
    let deleted: Bool
    let deletedAtMs: Int64?
    let nonce: Data
    let ciphertext: Data

    init(id: UUID, recordType: RecordType, keyVersion: Int, revision: HybridRevision, deviceID: UUID, updatedAtMs: Int64, deleted: Bool = false, deletedAtMs: Int64? = nil, nonce: Data, ciphertext: Data) {
        precondition(keyVersion > 0)
        precondition(!deleted || deletedAtMs != nil)
        self.id = id; self.recordType = recordType; schemaVersion = 1; self.keyVersion = keyVersion
        self.revision = revision; self.deviceID = deviceID; self.updatedAtMs = updatedAtMs
        self.deleted = deleted; self.deletedAtMs = deletedAtMs; self.nonce = nonce; self.ciphertext = ciphertext
    }

    var authenticatedMetadata: Data {
        Data("\(id.uuidString.lowercased())|\(recordType.rawValue)|\(schemaVersion)|\(keyVersion)|\(revision.encoded)|\(deleted)".utf8)
    }
}

File: app/src/main/java/com/example/sync/EncryptedRecordEnvelope.kt
package com.example.sync

import java.util.UUID

/** Metadata is intentionally the only data sent to Firestore in plaintext. */
enum class RecordType {
    PERSON, GROUP, GROUP_MEMBERSHIP, RELATIONSHIP, EVENT, EVENT_ATTENDEE,
    MEMORY, CAPTURE, REMINDER, SETTINGS
}

data class HybridRevision(
    val counter: Long,
    val deviceId: String
) : Comparable<HybridRevision> {
    init {
        require(counter >= 0) { "Revision counter must be non-negative" }
        require(runCatching { UUID.fromString(deviceId) }.isSuccess) { "deviceId must be a UUID" }
    }

    override fun compareTo(other: HybridRevision): Int =
        compareValuesBy(this, other, HybridRevision::counter, HybridRevision::deviceId)

    override fun toString(): String = "$counter:$deviceId"

    companion object {
        fun parse(value: String): HybridRevision {
            val separator = value.indexOf(':')
            require(separator > 0) { "Invalid revision" }
            return HybridRevision(value.substring(0, separator).toLong(), value.substring(separator + 1))
        }
    }
}

data class EncryptedRecordEnvelope(
    val id: String,
    val recordType: RecordType,
    val schemaVersion: Int = 1,
    val keyVersion: Int,
    val revision: HybridRevision,
    val deviceId: String,
    val updatedAtMs: Long,
    val deleted: Boolean,
    val deletedAtMs: Long? = null,
    val nonceBase64: String,
    val ciphertextBase64: String
) {
    init {
        require(runCatching { UUID.fromString(id) }.isSuccess) { "id must be a UUID" }
        require(schemaVersion == 1) { "Unsupported envelope schema" }
        require(keyVersion > 0) { "keyVersion must be positive" }
        require(updatedAtMs >= 0) { "updatedAtMs must be non-negative" }
        require(!deleted || deletedAtMs != null) { "Tombstones require a deletion timestamp" }
    }
}
